test/core/event_engine/test_suite/server_test.cc

```diff
@@ -12,12 +12,263 @@
+#include "src/core/lib/resource_quota/memory_quota.h"
+#include "src/core/lib/resource_quota/resource_quota.h"
 #include "test/core/event_engine/test_suite/event_engine_test.h"
+#include "test/core/event_engine/test_suite/event_engine_test_utils.h"
+#include "test/core/util/port.h"
 
 class EventEngineServerTest : public EventEngineTest {};
 
-// TODO(hork): establish meaningful tests
-TEST_F(EventEngineServerTest, TODO) { grpc_core::ExecCtx exec_ctx; }
+using namespace std::chrono_literals;
+
+namespace {
+
+using ::grpc_event_engine::experimental::ChannelArgsEndpointConfig;
+using ::grpc_event_engine::experimental::EventEngine;
+using ::grpc_event_engine::experimental::URIToResolvedAddress;
+using Endpoint = ::grpc_event_engine::experimental::EventEngine::Endpoint;
+using Listener = ::grpc_event_engine::experimental::EventEngine::Listener;
+using ::grpc_event_engine::experimental::GetNextSendMessage;
+using ::grpc_event_engine::experimental::WaitForSingleOwner;
+
+constexpr int kNumExchangedMessages = 100;
+
+}  // namespace
+
+// Create a connection using the oracle EventEngine to a listener created
+// by the Test EventEngine and exchange bi-di data over the connection.
+// For each data transfer, verify that data written at one end of the stream
+// equals data read at the other end of the stream.
+TEST_F(EventEngineServerTest, ServerConnectExchangeBidiDataTransferTest) {
+  grpc_core::ExecCtx ctx;
+  auto oracle_ee = this->NewOracleEventEngine();
+  std::shared_ptr<EventEngine> test_ee(this->NewEventEngine());
+  auto memory_quota = std::make_unique<grpc_core::MemoryQuota>("bar");
+  std::string target_addr = absl::StrCat(
+      "ipv6:[::1]:", std::to_string(grpc_pick_unused_port_or_die()));
+  std::unique_ptr<EventEngine::Endpoint> client_endpoint;
+  std::unique_ptr<EventEngine::Endpoint> server_endpoint;
+  grpc_core::Notification client_signal;
+  grpc_core::Notification server_signal;
+
+  Listener::AcceptCallback accept_cb =
+      [&server_endpoint, &server_signal](
+          std::unique_ptr<Endpoint> ep,
+          grpc_core::MemoryAllocator /*memory_allocator*/) {
+        server_endpoint = std::move(ep);
+        server_signal.Notify();
+      };
+
+  grpc_core::ChannelArgs args;
+  auto quota = grpc_core::ResourceQuota::Default();
+  args = args.Set(GRPC_ARG_RESOURCE_QUOTA, quota);
+  ChannelArgsEndpointConfig config(args);
+  auto status = test_ee->CreateListener(
+      std::move(accept_cb), [](absl::Status /*status*/) {}, config,
+      std::make_unique<grpc_core::MemoryQuota>("foo"));
+  EXPECT_TRUE(status.ok());
+
+  std::unique_ptr<Listener> listener = std::move(*status);
+  EXPECT_TRUE(listener->Bind(URIToResolvedAddress(target_addr)).ok());
+  EXPECT_TRUE(listener->Start().ok());
+
+  oracle_ee->Connect(
+      [&client_endpoint,
+       &client_signal](absl::StatusOr<std::unique_ptr<Endpoint>> status) {
+        if (!status.ok()) {
+          gpr_log(GPR_ERROR, "Connect failed: %s",
```

Comment:
Can this be an assert instead?

Reply:
Yes. replaced in both client_test.cc and server_test.cc